pkg/collector/scheduler.go

```diff
@@ -0,0 +1,190 @@
+				errorStats.setRunError(id, err.Error())
+			} else {
+				stopped[id] = struct{}{}
+			}
+		}
+
+		// remove the entry from `configToChecks`
+		if len(stopped) == len(s.configToChecks[digest]) {
+			// we managed to stop all the checks for this config
+			delete(s.configToChecks, digest)
+		} else {
+			// keep the checks we failed to stop in `configToChecks`
+			dangling := []check.ID{}
+			for _, id := range s.configToChecks[digest] {
+				if _, found := stopped[id]; !found {
+					dangling = append(dangling, id)
+				}
+			}
+			s.configToChecks[digest] = dangling
+		}
+	}
+}
+
+// Stop handles clean stop of registered schedulers
+func (s *CheckScheduler) Stop() {
+	if s.collector != nil {
+		s.collector.Stop()
+	}
+}
+
+// AddLoader adds a new Loader that AutoConfig can use to load a check.
+func (s *CheckScheduler) AddLoader(loader check.Loader) {
+	for _, l := range s.loaders {
+		if l == loader {
+			log.Warnf("Loader %s was already added, skipping...", loader)
+			return
+		}
+	}
+	s.loaders = append(s.loaders, loader)
+}
+
+// getChecks takes a check configuration and returns a slice of Check instances
+// along with any error it might happen during the process
+func (s *CheckScheduler) getChecks(config integration.Config) ([]check.Check, error) {
+	for _, loader := range s.loaders {
+		res, err := loader.Load(config)
+		if err == nil {
+			log.Debugf("%v: successfully loaded check '%s'", loader, config.Name)
+			errorStats.removeLoaderErrors(config.Name)
+			return res, nil
+		}
+		// Check if some check instances were loaded correctly (can occur if there's multiple check instances)
+		if len(res) != 0 {
+			return res, nil
+		}
+		errorStats.setLoaderError(config.Name, fmt.Sprintf("%v", loader), err.Error())
+		log.Debugf("%v: unable to load the check '%s': %s", loader, config.Name, err)
+	}
+
+	return []check.Check{}, fmt.Errorf("unable to load any check from config '%s'", config.Name)
+}
+
+// GetChecksFromConfigs gets all the check instances for given configurations
+// optionally can populate ac cache configToChecks
+func (s *CheckScheduler) GetChecksFromConfigs(configs []integration.Config, populateCache bool) []check.Check {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	var allChecks []check.Check
+	for _, config := range configs {
+		if !isCheckConfig(config) {
+			// skip non check configs.
+			continue
+		}
+		configDigest := config.Digest()
+		checks, err := s.getChecks(config)
+		if err != nil {
+			log.Errorf("Unable to load the check: %v", err)
+			continue
+		}
+		for _, c := range checks {
+			allChecks = append(allChecks, c)
+			if populateCache {
+				// store the checks we schedule for this config locally
+				s.configToChecks[configDigest] = append(s.configToChecks[configDigest], c.ID())
+			}
+		}
+	}
+
+	return allChecks
+}
+
+// isCheckConfig returns true if the config is a check configuration,
+func isCheckConfig(config integration.Config) bool {
+	return config.MetricConfig != nil || len(config.Instances) > 0
```

Comment:
Q: Does `config.MetricConfig != nil` actually qualify a check config? IIRC these are the default jmx metric listings, that the AD then appends to matching configs, right? These should not be actionable by themselves by the check scheduler

Reply:
I agree, @ajacquemot do you recall why you added this condition?

Reply:
You might just have a `LogsConfig` defined in your `config` and there is not point to process it here, there might be a better condition for sure.